app/models/user.rb

```diff
@@ -1,98 +1,30 @@
 class User < ApplicationRecord
+  PROPERTIES = %i(name email address phone password
+                  password_confirmation remember_me).freeze
+  devise :database_authenticatable, :registerable, :validatable,
```

Comment:
Làm sao để anh có thể thay đổi độ phức tạp của mật khẩu em nhỉ?
làm sao mà devise chỉ lưu mỗi thời gian remember mà vẫn biết được đấu là người dùng đúng?

Reply:
trong file confing/initializers/devise.rb chúng ta có thể config mộ số option cho devise. muốn thay đổi độ phức tạp thì ta sửa dòng   
# config.password_length = 8..72
hoặc ta có thể bỏ option validatable và tự validate trong model ạ

Reply:
User.find(1).remember_me!  # regenerating the token
User.find(1).forget_me!    # clearing the token

# generating info to put into cookies
User.serialize_into_cookie(user)

# lookup the user based on the incoming cookie information
User.serialize_from_cookie(cookie_string)
khi click nút remember me thì devise sẽ sinh ra cookie bên trong có chứa remember_digist được hash và trả về cho client mã này cũng đồng thời lưu trong trường remember_digist của User trong db. Trong những lần đăng nhập tiếp theo thì client gửi kèm cookie lên server.Nếu cookie hợp lệ Devise sẽ tìm trong db so sánh với cookie để nhận dạng xem đó là User nào ạ